### PR TITLE
Shopping cart implementation

### PR DESCRIPTION
**Pull request by mkazarian-1**

### CODE REVIEW

File: src/main/java/org/example/bookshop/controller/ShoppingCartController.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    @Operation(summary = "Create book",
+            description = """
+                    Create cart item object and Returns shopping cart date with
+                    list of all available cart items
+                    Throws DuplicateException if the
+                    cart item with entered book already exists.
+                    \nNecessary role: USER
+                    """)
+    public ShoppingCartDto saveItem(@RequestBody @Valid CreateCartItemRequestDto cartItemRequestDto,
+                                    @PageableDefault(size = 5) Pageable pageable) {
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/controller/ShoppingCartController.java]:
```suggestion
                                   ) {
```

File: src/main/java/org/example/bookshop/controller/ShoppingCartController.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    @PreAuthorize("hasAuthority('USER')")
+    @Operation(summary = "Get shopping cart (with pagination and sorting for shopping cart)",
+            description = """
+                    Returns shopping cart date with
+                    list of all available cart items
+                    by pages and give ability
+                    to sort cart items according to the specified parameters
+                    \nNecessary role: USER
+                    """)
+    public ShoppingCartDto getShoppingCart(@PageableDefault(size = 5) Pageable pageable) {
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/controller/ShoppingCartController.java]:
```suggestion
    public ShoppingCartDto getShoppingCart() {
```

File: src/main/java/org/example/bookshop/controller/ShoppingCartController.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+            description = """
+                    Return the update cart item if the update went well.
+                    Throws DuplicateException
+                    if the cart item with entered book already exists.
+                    \nNecessary role: USER
+                    """)
+    public ShoppingCartDto updateCartItem(
+            @RequestBody @Valid UpdateCartItemRequestDto cartItemRequestDto,
+                                          @PathVariable Long id,
+                                          @PageableDefault(size = 5) Pageable pageable) {
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/controller/ShoppingCartController.java]:
```suggestion
                                          ) {
```

File: src/main/java/org/example/bookshop/dto/cart/item/CreateCartItemRequestDto.java [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.NotNull;
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+public class CreateCartItemRequestDto {
+    @NotNull
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/dto/cart/item/CreateCartItemRequestDto.java:11]:
```suggestion
    @NotNull
    @Positive
```

File: src/main/java/org/example/bookshop/dto/cart/item/CreateCartItemRequestDto.java [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+    private Long bookId;
+    @Min(value = 1)
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/dto/cart/item/CreateCartItemRequestDto.java]:
```suggestion
    @Positive
```

File: src/main/java/org/example/bookshop/dto/cart/item/UpdateCartItemRequestDto.java [hunk 6]
```diff
@@ -0,0 +1,12 @@
+package org.example.bookshop.dto.cart.item;
+
+import jakarta.validation.constraints.Min;
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+public class UpdateCartItemRequestDto {
+    @Min(value = 0)
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/dto/cart/item/UpdateCartItemRequestDto.java]:
```suggestion
    @Positive
```

File: src/main/java/org/example/bookshop/mapper/BookMapper.java [hunk 7]
```diff
@@ -53,4 +53,13 @@ default Set<Category> setCategoriesFromIds(List<Long> categoryIds) {
...
     BookDtoWithoutCategoryIds toDtoWithOutCategoryIds(Book book);
+
+    @Named("bookFromId")
+    default Book bookFromId(Long bookId) {
+        Book book = new Book();
+        book.setId(bookId);
+        return book;
+    }
 }
+
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/mapper/BookMapper.java]: remove redundant empty lines

File: src/main/java/org/example/bookshop/model/CartItem.java [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+@Getter
+@Setter
+@Table(name = "cart_items")
+public class CartItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @ManyToOne
+    @JoinColumn(name = "shopping_cart_id", nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/model/CartItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "shopping_cart_id", nullable = false)
```

File: src/main/java/org/example/bookshop/model/CartItem.java [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    private ShoppingCart shoppingCart;
+
+    @ManyToOne
+    @JoinColumn(name = "book_id", nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/model/CartItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "book_id", nullable = false)
```

File: src/main/java/org/example/bookshop/service/impl/ShoppingCartManagerServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final CartItemRepository cartItemRepository;
+    private final CartItemMapper cartItemMapper;
+    private final ShoppingCartService shoppingCartService;
+    private final BookRepository bookRepository;
+
+    @Override
+    public ShoppingCartDto saveCartItem(
+            Pageable pageable, CreateCartItemRequestDto cartItemRequestDto) {
+        if (!bookRepository.existsById(cartItemRequestDto.getBookId())) {
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/service/impl/ShoppingCartManagerServiceImpl.java]: see the example of implementation
```suggestion
        Book book = bookRepository.findById(itemDto.bookId())
                .orElseThrow(() -> new EntityNotFoundException(
                        String.format("Book with id: %d is not found", itemDto.bookId())
                ));
        ShoppingCart cart = repository.findByUserId(userId);
        cart.getCartItems().stream()
                .filter(item -> item.getBook().getId().equals(itemDto.bookId()))
                .findFirst()
                .ifPresentOrElse(item -> item.setQuantity(item.getQuantity() + itemDto.quantity()),
                        () -> addCartItemToCart(itemDto, book, cart));
        repository.save(cart);
        return cartMapper.toDto(cart);
```

File: src/main/java/org/example/bookshop/service/impl/ShoppingCartManagerServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+
+        CartItem cartItem = cartItemMapper.toEntity(cartItemRequestDto);
+        cartItem.setShoppingCart(shoppingCart);
+        cartItemRepository.save(cartItem);
+
+        return getShoppingCart(pageable);
+    }
+
+    @Override
+    public ShoppingCartDto getShoppingCart(Pageable pageable) {
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/service/impl/ShoppingCartManagerServiceImpl.java]:
```suggestion
    public ShoppingCartDto getShoppingCart(Long userId) {
```

File: src/main/java/org/example/bookshop/service/impl/ShoppingCartManagerServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+        ShoppingCart shoppingCart = shoppingCartService.getShoppingCartByCurrentUser();
+        shoppingCart.setCartItems(new HashSet<>(cartItemRepository
+                .findCartItemsByShoppingCart(shoppingCart, pageable)));
+
+        return shoppingCartMapper.toDto(shoppingCart);
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/service/impl/ShoppingCartManagerServiceImpl.java]:
```suggestion
        return cartMapper.toDto(repository.findByUserId(userId));
```

File: src/main/java/org/example/bookshop/service/impl/ShoppingCartManagerServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+        CartItem cartItem = cartItemRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "There is no cartItem with the specified ID:" + id));
+
+        cartItemMapper.updateCartItemFromDto(
+                cartItem, updateCartItemRequestDto
+        );
+        cartItemRepository.save(cartItem);
+
+        return getShoppingCart(pageable);
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/service/impl/ShoppingCartManagerServiceImpl.java]:
```suggestion
        ShoppingCart cart = repository.findByUserId(userId);
        CartItem cartItem = itemRepository.findByIdAndShoppingCartId(cartItemId, cart.getId())
                .map(item -> {
                    item.setQuantity(itemDto.quantity());
                    return item;
                }).orElseThrow(() -> new EntityNotFoundException(
                        String.format("No cart item with id: %d for user: %d", cartItemId, userId)
                ));
        itemRepository.save(cartItem);
        return cartMapper.toDto(cart);
```

File: src/main/java/org/example/bookshop/service/impl/ShoppingCartServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+import org.example.bookshop.model.ShoppingCart;
+import org.example.bookshop.model.User;
+import org.example.bookshop.repository.ShoppingCartRepository;
+import org.example.bookshop.service.ShoppingCartService;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class ShoppingCartServiceImpl implements ShoppingCartService {
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/service/impl/ShoppingCartServiceImpl.java:22]: remove this service, use only ShopCartService

File: src/main/java/org/example/bookshop/service/impl/ShoppingCartServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+    private final ShoppingCartRepository shoppingCartRepository;
+
+    @Override
+    public void createShoppingCart(User user) {
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/service/impl/ShoppingCartServiceImpl.java:30]: replace this method to ShoppingCartServiceImpl

File: src/main/resources/db/changelog/changes/08-insert-shopping-cart-for-admin-user.yaml [hunk 16]
```diff
@@ -0,0 +1,17 @@
+databaseChangeLog:
+  - changeSet:
+      id: 1
+      author: your_name
+      changes:
+        - insert:
+            tableName: shopping_carts
+            columns:
+              - column:
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/08-insert-shopping-cart-for-admin-user.yaml:9]: why do you need this changelog?

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/resources/db/changelog/changes/08-insert-shopping-cart-for-admin-user.yaml (hunk 16 above)
Comment (mkazarian-1) [reply src/main/resources/db/changelog/changes/08-insert-shopping-cart-for-admin-user.yaml:9]: As we added admin user manually we will get exception if admin decide to use shop cart

File: src/main/java/org/example/bookshop/dto/cart/item/CreateCartItemRequestDto.java (hunk 4 above)
Comment (ahoienko) [reply src/main/java/org/example/bookshop/dto/cart/item/CreateCartItemRequestDto.java:11]: Not resolved

File: src/main/java/org/example/bookshop/service/impl/ShoppingCartServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+
+        shoppingCartRepository.save(cart);
+
+        return shoppingCartMapper.toDto(cart);
+    }
+
+    @Override
+    public ShoppingCartDto getShoppingCart(Long userId) {
+        return shoppingCartMapper.toDto(
+                shoppingCartRepository.getByUserId(userId)
```
Comment (ahoienko) [inline src/main/java/org/example/bookshop/service/impl/ShoppingCartServiceImpl.java]: What if shopping cart is missing?

File: src/main/java/org/example/bookshop/repository/ShoppingCartRepository.java [hunk 18]
```diff
@@ -0,0 +1,10 @@
+package org.example.bookshop.repository;
+
+import org.example.bookshop.model.ShoppingCart;
+import org.springframework.data.jpa.repository.EntityGraph;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface ShoppingCartRepository extends JpaRepository<ShoppingCart, Long> {
+    @EntityGraph(attributePaths = {"cartItems","cartItems.book","user"})
+    ShoppingCart getByUserId(Long id);
```
Comment (ahoienko) [inline src/main/java/org/example/bookshop/repository/ShoppingCartRepository.java]:
```suggestion
    Optional<ShoppingCart> getByUserId(Long id);
```

**Review by ahoienko — CHANGES_REQUESTED** [review]
Nice, left minor comments

File: src/main/java/org/example/bookshop/controller/ShoppingCartController.java [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    @PutMapping("items/{cartItemId}")
+    @PreAuthorize("hasAuthority('USER')")
+    @Operation(summary = "Update shopping cart",
+            description = """
+                    Return the update cart item if the update went well.
+                    \nNecessary role: USER
+                    """)
+    public CartItemDto updateCartItem(
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/controller/ShoppingCartController.java]:
```suggestion
    public ShoppingCartDto updateCartItem(
```

File: src/main/java/org/example/bookshop/dto/cart/item/CreateCartItemRequestDto.java (hunk 4 above)
Comment (Elena-Bruyako) [reply src/main/java/org/example/bookshop/dto/cart/item/CreateCartItemRequestDto.java:11]: Add Positive 
```suggestion
    @NotNull
    @Positive
```

File: src/main/java/org/example/bookshop/dto/cart/item/CreateCartItemRequestDto.java [hunk 20]
```diff
@@ -0,0 +1,16 @@
...
+import jakarta.validation.constraints.Positive;
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+public class CreateCartItemRequestDto {
+    @NotNull
+    private Long bookId;
+    @NotNull
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/dto/cart/item/CreateCartItemRequestDto.java]: int type can't be null
remove annotation
```suggestion
```

File: src/main/java/org/example/bookshop/dto/cart/item/UpdateCartItemRequestDto.java [hunk 21]
```diff
@@ -0,0 +1,14 @@
...
+
+import jakarta.validation.constraints.NotNull;
+import jakarta.validation.constraints.Positive;
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+public class UpdateCartItemRequestDto {
+    @NotNull
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/dto/cart/item/UpdateCartItemRequestDto.java]: same here
```suggestion
```

File: src/main/java/org/example/bookshop/service/ShoppingCartService.java [hunk 22]
```diff
@@ -0,0 +1,23 @@
...
+import org.example.bookshop.model.User;
+
+public interface ShoppingCartService {
+    void createShoppingCart(User user);
+
+    ShoppingCartDto saveCartItem(CreateCartItemRequestDto cartItemRequestDto, Long userId);
+
+    ShoppingCartDto getShoppingCart(Long userId);
+
+    CartItemDto updateCartItem(
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/service/ShoppingCartService.java]:
```suggestion
    ShoppingCartDto updateCartItem(
```

File: src/main/java/org/example/bookshop/service/impl/ShoppingCartServiceImpl.java [hunk 23]
```diff
@@ -0,0 +1,109 @@
...
+    private final BookRepository bookRepository;
+
+    @Override
+    public void createShoppingCart(User user) {
+        ShoppingCart shoppingCart = new ShoppingCart();
+        shoppingCart.setUser(user);
+        shoppingCartRepository.save(shoppingCart);
+    }
+
+    @Override
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/service/impl/ShoppingCartServiceImpl.java:36]: Do we need `@Transactional` here?

File: src/main/java/org/example/bookshop/service/impl/ShoppingCartServiceImpl.java [hunk 24]
```diff
@@ -0,0 +1,109 @@
...
+                        .getId()
+                        .equals(cartItemRequestDto.getBookId()))
+                .findFirst()
+                .ifPresentOrElse(item -> item.setQuantity(item.getQuantity()
+                                + cartItemRequestDto.getQuantity()),
+                        () -> addCartItemToCart(
+                                cartItemRequestDto, cart, book));
+
+        shoppingCartRepository.save(cart);
+
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/service/impl/ShoppingCartServiceImpl.java]: remove redundant empty line
```suggestion
```

File: src/main/java/org/example/bookshop/service/impl/ShoppingCartServiceImpl.java [hunk 25]
```diff
@@ -0,0 +1,109 @@
...
+        cart.getCartItems().stream()
+                .filter(item -> item.getBook()
+                        .getId()
+                        .equals(cartItemRequestDto.getBookId()))
+                .findFirst()
+                .ifPresentOrElse(item -> item.setQuantity(item.getQuantity()
+                                + cartItemRequestDto.getQuantity()),
+                        () -> addCartItemToCart(
+                                cartItemRequestDto, cart, book));
+
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/service/impl/ShoppingCartServiceImpl.java:57]: remove redundant empty line

File: src/main/java/org/example/bookshop/service/impl/ShoppingCartServiceImpl.java [hunk 26]
```diff
@@ -0,0 +1,109 @@
...
+    public ShoppingCartDto getShoppingCart(Long userId) {
+        return shoppingCartMapper.toDto(
+                shoppingCartRepository.findByUserId(userId)
+                        .orElseThrow(() -> new EntityNotFoundException(String.format(
+                                "Can't find Shopping cart with user id: %d", userId)))
+        );
+    }
+
+    @Override
+    public CartItemDto updateCartItem(
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
    public ShoppingCartDto updateCartItem(
```

File: src/main/java/org/example/bookshop/service/impl/UserServiceImpl.java [hunk 27]
```diff
@@ -36,8 +38,11 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto) {
 
         User user = userMapper.toUser(requestDto);
         userMapper.setPasswordAndRole(user, requestDto, passwordEncoder, roleRepository);
+        user = userRepository.save(user);
 
-        return userMapper.toDto(userRepository.save(user));
+        shoppingCartService.createShoppingCart(user);
+
+        return userMapper.toDto(user);
     }
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/service/impl/UserServiceImpl.java:50]: Why login methor is here? It should be in AuthenticationService

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/resources/db/changelog/changes/08-insert-shopping-cart-for-admin-user.yaml [hunk 28]
```diff
@@ -0,0 +1,17 @@
+databaseChangeLog:
+  - changeSet:
+      id: 1
+      author: your_name
```
Comment (okuzan) [inline src/main/resources/db/changelog/changes/08-insert-shopping-cart-for-admin-user.yaml]:
```suggestion
      author: mkUser
```

File: src/main/java/org/example/bookshop/service/impl/ShoppingCartServiceImpl.java [hunk 29]
```diff
@@ -0,0 +1,104 @@
...
+        shoppingCartRepository.save(cart);
+        return getShoppingCart(userId);
+    }
+
+    @Override
+    public ShoppingCartDto getShoppingCart(Long userId) {
+        return shoppingCartMapper.toDto(
+                shoppingCartRepository.findByUserId(userId)
+                        .orElseThrow(() -> new EntityNotFoundException(String.format(
+                                "Can't find Shopping cart with user id: %d", userId)))
```
Comment (okuzan) [inline src/main/java/org/example/bookshop/service/impl/ShoppingCartServiceImpl.java]: repeated thrice, let's extract to a private method

**Review by okuzan — CHANGES_REQUESTED** [review]

**Pull request merged**